### PR TITLE
Require a semaphore in Consumer.fromJavaConsumer

### LINKED ISSUE #1276: FromJavaConsumer: how to synchronize access to the java consumer
#694 introduced a `fromJavaConsumer` constructor. But it isn't clear how to synchronize access to the consumer once handed to zio-kafka. My use case is the same as the one reported in #694. I have some logic working on the java consumer to check for the consumer lag, health and other metrics. These checks run periodically but I need to coordinate the accesses with zio-kafka. As far as I understand, zio-kafka doens't provide that ability and I'll face concurrency issues if both my checks and zio-kafka run tasks against the consumer concurrently.

### PR DESCRIPTION
**Pull request by svroonland**
Fixes #1276

### CODE REVIEW

File: zio-kafka/src/main/scala/zio/kafka/consumer/Consumer.scala [hunk 1]
```diff
@@ -199,15 +199,44 @@ object Consumer {
...
+  def fromJavaConsumerWithPermit(
+    javaConsumer: JConsumer[Array[Byte], Array[Byte]],
+    settings: ConsumerSettings,
+    permit: Semaphore,
+    diagnostics: Diagnostics = Diagnostics.NoOp
+  ): ZIO[Scope, Throwable, Consumer] =
+    for {
+      _ <- ZIO.addFinalizer(diagnostics.emit(Finalization.ConsumerFinalized))
+      consumerAccess = new ConsumerAccess(javaConsumer, permit)
+      runloopAccess <- RunloopAccess.make(settings, consumerAccess, diagnostics)
```
Comment (erikvanoosten) [inline zio-kafka/src/main/scala/zio/kafka/consumer/Consumer.scala:239]:
```suggestion
  /**
   * Create a zio-kafka [[Consumer]] from an `org.apache.kafka KafkaConsumer`.
   *
   * You are responsible for all of the following:
   *  - creating and closing the `KafkaConsumer`,
   *  - making sure `auto.commit` is disabled,
   *  - creating `access` as a fair semaphore with a single permit,
   *  - acquire a permit from `access` before using the consumer, and release if afterwards,
   *  - not using the following consumer methods: `subscribe`, `unsubscribe`, `assign`, `poll`, `commit*`, `seek`, `pause`, `resume`, and `enforceRebalance`.
   *
   * Any deviation of these rules is likely to cause hard to track errors.
   *
   * Semaphore `access` is shared between you and the zio-kafka consumer. Use it as short as possible; while you hold a permit the zio-kafka consumer is blocked.
   *
   * @param javaConsumer
   *   Consumer
   * @param settings
   *   Settings
   * @param access
   *   A Semaphore with 1 permit.
   * @param diagnostics
   *   Optional diagnostics listener
   */
  def fromJavaConsumerWithPermit(
    javaConsumer: JConsumer[Array[Byte], Array[Byte]],
    settings: ConsumerSettings,
    access: Semaphore,
    diagnostics: Diagnostics = Diagnostics.NoOp
  ): ZIO[Scope, Throwable, Consumer] =
    for {
      _ <- ZIO.addFinalizer(diagnostics.emit(Finalization.ConsumerFinalized))
      consumerAccess = new ConsumerAccess(javaConsumer, access)
      runloopAccess <- RunloopAccess.make(settings, consumerAccess, diagnostics)
```

**Review by erikvanoosten — APPROVED** [review]
A very pragmatic solution, let's ship it. Some doc improvement suggested.

File: zio-kafka-test/src/test/scala/zio/kafka/consumer/internal/RunloopSpec.scala [hunk 2]
```diff
@@ -195,8 +195,9 @@ object RunloopSpec extends ZIOSpecDefaultSlf4j {
   ): ZIO[Scope, Throwable, TestResult] =
     ZIO.scoped {
       for {
-        consumerAccess <- ConsumerAccess.make(mockConsumer)
-        consumerScope  <- ZIO.scope
+        permit <- Semaphore.make(1)
+        consumerAccess = new ConsumerAccess(mockConsumer, permit)
```
[the remark this answers is not in the document]
Comment (svroonland) [reply zio-kafka-test/src/test/scala/zio/kafka/consumer/internal/RunloopSpec.scala:199]: Ah that would be more consistent, although I'm not sure I like the parameter name `access` in the first place.

[the remark this answers is not in the document]
Comment (erikvanoosten) [reply zio-kafka-test/src/test/scala/zio/kafka/consumer/internal/RunloopSpec.scala]: I agree, its not great. But `permit` is plain wrong 😉

**Pull request merged**